### DISCUSSION TITLE
WASM - Unable to detect Ctrl+D, browser opens add favorite, but can capture in JavaScript

### DISCUSSION THREAD

**Discussion by darenm** [unoplatform/uno#11040] [Q&A]
Hi, I am trying to attach an operation to CTRL+D key combination in a UWP/WASM project. It works fine in UWP (as expected), however the browser is capturing the keypress and displaying the **Add/Edit Favorites** UI in the WASM head. 

Using javascript I can capture the keypresses using something like:

```javascript
document.addEventListener('keydown', function(event) {
  if (event.ctrlKey && event.key === 'd') {
    alert('CTRL+D');
    event.preventDefault(); // stops the favorites UI from receiving the event
  }
});
```

See [jsfiddle](https://jsfiddle.net/1c9y6r7q/).

Is there a way to configure Uno to support this?

**Comment by Rytisgit** [comment]
Yes, but not in the C# code. You would need to use JavaScript and possibly callback to the C# code by binding the static method: https://platform.uno/docs/articles/interop/wasm-javascript-1.html#invoke-c-code-from-javascript

The way to embed the JavaScript code is explained in the same page.

**Reply by darenm** [reply]
Thanks, I'm familiar with that approach - I was hoping for some configuration option that could enable it for capture using the "regular" approach. :) 

Similarly, I have the following code that allows me to capture CTRL+MouseWheel so that I can zoom documents within the WASM app, rather than zoom the entire browser:

```javascript
function stopWheel(e) {
  if (!e) {
    e = window.event;
  } /* IE7, IE8, Chrome, Safari */
  if (e.preventDefault) {
    e.preventDefault();
  } /* Chrome, Safari, Firefox */
  e.returnValue = false; /* IE7, IE8 */
}

function detectMouseWheelDirection(e) {
  var delta = null,
    direction = false;
  if (!e) { // if the event is not provided, we get it from the window object       
    e = window.event;
  }
  if (e.wheelDelta) { // will work in most cases        
    delta = e.wheelDelta / 60;
  } else if (e.detail) { // fallback for Firefox        
    delta = -e.detail / 2;
  }
  if (delta !== null) {
    direction = delta > 0 ? 'up' : 'down';
  }
  return direction;
}

document.addEventListener('mousewheel', function(event) {
  if (event.ctrlKey) {
    stopWheel();
    var direction = detectMouseWheelDirection(event);
    console.log(direction);
  }
}, {
  passive: false
});
if (document.addEventListener) {
  /* Chrome, Safari, Firefox */
  document.addEventListener('DOMMouseScroll', stopWheel, {
    passive: false
  });
}
```

**Reply by jeromelaban** [reply]
You'll notice that the [uno calculator handles it properly](https://calculator.platform.uno) and there's no javascript there, which means that hooking onto the Window.KeyDown is helping in that direction.

**Reply by darenm** [reply]
I see, so using CoreWindow is the path:

```csharp
var coreWindow = Window.Current.Content;
try
{
	coreWindow.KeyDown += (snd, e) =>
	{
		target.OnKeyPress(e.Key);
		e.Handled = true;
	};
}
catch (Exception)
{
}
```

**Reply by Rytisgit** [reply]
There is still the issue of tracking if the control key is pressed or not, because as far as I know there's no way to see if a key is held down, it is only possible to be handled by chording. This doesn't account for a user coming into focus or leaving focus from your application with CTRL held down. Currently all the calculator app basically does is set e.Handled for the CTRL keypress so that it doesn't activate the bookmark in the browser, but doesn't do anything with the data additionally. I believe it has the same issues with focus I described from what I looked at the code.
